### PR TITLE
Use bitset in MakePredicatorIsReachable to reduce memory usage

### PR DESCRIPTION
**Pull request by jackalcooper**

### CODE REVIEW

File: oneflow/core/graph/graph.h [hunk 1]
```diff
@@ -608,26 +609,52 @@ Graph<NodeType, EdgeType>::MakePredicatorIsReachable() const {
...
+    (*node2ancestor)[node].resize(bitset_num);
+    id += 1;
+  });
   TopoForEachNode(starts, ForEachInNode, ForEachOutNode, [&](NodeType* node) {
-    node2ancestor->emplace(node, HashSet<const NodeType*>());
     ForEachInNode(node, [&](NodeType* in_node) {
-      node2ancestor->at(node).insert(in_node);
-      node2ancestor->at(node).insert(node2ancestor->at(in_node).begin(),
-                                     node2ancestor->at(in_node).end());
+      const int64_t in_node_id = (*node2id).at(in_node);
```
Comment (liujuncheng) [inline oneflow/core/graph/graph.h:650]:
```suggestion
      const int64_t in_node_id = node2id->at(in_node);
```

File: oneflow/core/graph/graph.h [hunk 2]
```diff
@@ -608,26 +609,52 @@ Graph<NodeType, EdgeType>::MakePredicatorIsReachable() const {
...
 template<typename NodeType, typename EdgeType>
 std::function<bool(const NodeType* src, const NodeType* dst)>
 Graph<NodeType, EdgeType>::MakePredicatorIsReachable(
     const std::list<NodeType*>& starts,
     const std::function<void(NodeType*, const std::function<void(NodeType*)>&)>& ForEachInNode,
     const std::function<void(NodeType*, const std::function<void(NodeType*)>&)>& ForEachOutNode)
     const {
-  auto node2ancestor = std::make_shared<HashMap<const NodeType*, HashSet<const NodeType*>>>();
+  auto node2ancestor = std::make_shared<HashMap<const NodeType*, bitset_vec>>();
+  auto node2id = std::make_shared<HashMap<const NodeType*, int64_t>>();
```
Comment (liujuncheng) [inline oneflow/core/graph/graph.h:640]: node2id->reserve(num_nodes());

File: oneflow/core/graph/graph.h [hunk 3]
```diff
@@ -615,19 +616,56 @@ Graph<NodeType, EdgeType>::MakePredicatorIsReachable(
...
+  });
   TopoForEachNode(starts, ForEachInNode, ForEachOutNode, [&](NodeType* node) {
-    node2ancestor->emplace(node, HashSet<const NodeType*>());
+    const int64_t node_id = node2id->at(node);
     ForEachInNode(node, [&](NodeType* in_node) {
-      node2ancestor->at(node).insert(in_node);
-      node2ancestor->at(node).insert(node2ancestor->at(in_node).begin(),
-                                     node2ancestor->at(in_node).end());
+      const int64_t in_node_id = node2id->at(in_node);
+      auto& ancestor_bitset_vec = id2ancestor->at(node_id);
```
Comment (liujuncheng) [inline oneflow/core/graph/graph.h:661]: 这一样提到`ForEachInNode`外面吧

Comment (jackalcooper) [reply oneflow/core/graph/graph.h]: 已改

**Review by liujuncheng — APPROVED** [review]

**Pull request merged**